### ISSUE TITLE
processArguments not being sent to app iOS10

### ISSUE THREAD

**Issue by kalawton** [appium/appium#6888]
## The problem

I am using processArguments, but the app is not getting that information on startup.
## Environment
- Appium version (or git revision) that exhibits the issue: Appium v1.6.0-beta2 (REV 9441a89f7e368f0c621db3a0cf9ab074c045ac3a)
- Last Appium version that did not exhibit the issue (if applicable): 1.4
- Desktop OS/version used to run Appium: MacOS 10.11.6
- Node.js version: 5.4.1
- Mobile platform/version under test: iOS10 iPhone 6s
- Real device or emulator/simulator: Simulator
- Appium CLI or Appium.app|exe:  CLI
## Details

The emulator does start, and the app launches, but the app logs tell me that the launchArguments are not being passed to the app.  I am not sure if the arguments are being passed by WDA or not.
## Link to Appium logs

https://gist.github.com/kalawton/c89ba1428e8923b8f4032a227b41cf0d
## Code To Reproduce Issue [ Good To Have ]

`
String[] baseURL = {"-baseURL "+MobileProperties.getProperty("selected.environment")};
        JSONObject arguments = new JSONObject();
        arguments.put("args", baseURL);

desiredCapabilities capabilities = new DesiredCapabilities();
        capabilities.setCapability("automationName", "XCUITest");
        capabilities.setCapability("platformName", "iOS");
        capabilities.setCapability("platformVersion", iOS_VERSION);
        capabilities.setCapability("deviceName", iOS_DEVICE);
        capabilities.setCapability("app", appPath);
        capabilities.setCapability("autoLaunch", true);
        capabilities.setCapability("autoAcceptAlerts", true);
        capabilities.setCapability("sendKeyStrategy", "grouped");
        capabilities.setCapability("processArguments", arguments) ;
        capabilities.setCapability("language", iOS_Language);
        capabilities.setCapability("locale", iOS_REGION);
        capabilities.setCapability("fullReset", true);
`

**Comment by mykola-mokhnach** [comment]
I do it like this and it works so far:

```java
        final List<String> processArgs = new ArrayList<>(Arrays.asList(
                "--disable-autocorrection"
                ,"--debug-log-network"
        ));
        final JSONObject argsValue = new JSONObject();
        argsValue.put("args", processArgs);
        capabilities.setCapability("processArguments", argsValue.toString());
```

**Comment by kalawton** [comment]
@mykola-mokhnach your form does work. Thank you.  

I notice that this is adding escape characters to the desiredCapabilities being used.  I was not expecting to need all those escapes.

`POST /wd/hub/session {"desiredCapabilities":{"app":"/Users/kirtlawton/bowman/UI_Tests/src/main/resources/mobile/apps/RelayRides.app","language":"en","processArguments":"{\"args\":[\"-baseURL http:\\/\\/ec2-54-69-226-209.us-west-2.compute.amazonaws.com\"]}","locale":"en_US","deviceName":"iPhone 6 Plus","fullReset":true,"showXcodeLog":false,"sendKeyStrategy":"grouped","autoLaunch":true,"platformVersion":"10.0","automationName":"XCUITest","platformName":"iOS","autoAcceptAlerts":true}}`

**Comment by kalawton** [comment]
I switched up my code to match @mykola-mokhnach suggestion.  It has added escape sequences to the url, but the result is still the same. The simulator is not starting with the URL passed in by processArguments.

`List<String> baseURL = new ArrayList<>(Arrays.asList("-baseURL "+MobileProperties.getProperty("selected.environment")));
JSONObject arguments = new JSONObject();
arguments.put("args", baseURL);
capabilities.setCapability("processArguments", arguments.toString()) ;
`
**Output**
[Appium] Capabilities:
[Appium]   app: '/Users/kirtlawton/bowman/UI_Tests/src/main/resources/mobile/apps/RelayRides.app'
[Appium]   language: 'en'
[Appium]   processArguments: '{"args":["-baseURL http:\\/\\/ec2-54-69-226-209.us-west-2.com"]}'
[Appium]   locale: 'en_US'
[Appium]   deviceName: 'iPhone 6 Plus'
[Appium]   fullReset: true
[Appium]   showXcodeLog: false
[Appium]   sendKeyStrategy: 'grouped'
[Appium]   autoLaunch: true
[Appium]   platformVersion: '10.0'
[Appium]   automationName: 'XCUITest'
[Appium]   platformName: 'iOS'
[Appium]   autoAcceptAlerts: true
[BaseDriver] Session created with session id: 2404b23a-d50f-4b05-a7bb-5f5e6c985961

**Comment by mykola-mokhnach** [comment]
probably, in you case, there should be two separate args array items:

```java
List<String> baseURL = new ArrayList<>(Arrays.asList("-baseURL "+MobileProperties.getProperty("selected.environment"))); 
```

replace with

```java
List<String> baseURL = new ArrayList<>(Arrays.asList(
  "-baseURL",
  MobileProperties.getProperty("selected.environment")
)); 
```

**Comment by kalawton** [comment]
Makes sense, but this change had no effect.   Are there logs for WDA that I can get more information from?  I would like to verify what it is getting from the appium server.

**Comment by mykola-mokhnach** [comment]
I would say you should add more logging to your app under test to see what it receives as arguments.

**Comment by imurchie** [comment]
You can use the `showIosLog` capability to get _all_ the logs from the device.

**Comment by kalawton** [comment]
The app logs show that no arguments are being passed in during startup:
baseURLFromArguments: (null)

**Comment by imurchie** [comment]
Can you provide the logs so that we can look and see how to fix this issue? Our test cases work.

**Comment by kalawton** [comment]
Here is the Gist of that last run.
https://gist.github.com/kalawton/e66016b8c3985759621e4812aad7ce83

**Comment by yaphatak** [comment]
@kalawton I see this `Error: The desired capabilities must include either an app or a bundleId for iOS`. You need to pass either of those (app or a bundleId )

**Comment by kalawton** [comment]
@yaphatak ya, I can't explain that error, but I am sending the app capabilities.  This might have come in because I killed the test process to keep the logs from getting too long.  The problem I am facing happens towards the beginning of the test run.
The app does launch; it just is not launching with the correct arguments.

**Comment by codiophile** [comment]
Are there any news on this issue? This is currently blocking us from migrating to Xcode 8 and testing on iOS 10. I saw the merge request for WebDriverAgent and I'm wondering if anyone knows if this is a bug in WebDriverAgent or Appium?

I don't have much experience with WebDriverAgent or Appium, but if there is anything I can do to help out, please let me know. I am very keen on getting this resolved.

**Comment by codiophile** [comment]
While debugging this I realised that I was sending the processArguments in the wrong format. I fixed the format and now it is working. As far as I can tell the OP is using the correct format, so not sure why he is having this issue. The only thing I can see in the log is that there is an extra space that shouldn't be necessary: "-baseURL " should be "-baseURL", but I don't think an extra space should make any difference. Could be worth checking, I guess.
